app/src/main/java/com/android/sample/model/preferences/PreferencesLocalCache.kt

```diff
@@ -0,0 +1,87 @@
+package com.android.sample.model.preferences
+
+import android.content.Context
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.emptyPreferences
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.datastore.preferences.preferencesDataStore
+import com.google.gson.Gson
+import java.io.IOException
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.map
+
+val Context.preferencesDataStore: DataStore<Preferences> by
+    preferencesDataStore(name = "preferences_cache")
+
+class PreferencesLocalCache(private val context: Context) {
+  private val gson = Gson()
+  private val preferencesKey = stringPreferencesKey("preferences_data")
+
+  /**
+   * Retrieves the user preferences from local cache.
+   *
+   * @return A [Flow] of [Preferences] object.
+   */
+  fun getPreferences(): Flow<com.android.sample.model.preferences.Preferences?> =
+      context.preferencesDataStore.data
+          .catch { exception ->
+            when (exception) {
+              is IOException -> {
+                Log.e(
+                    "PreferencesLocalCache",
+                    "Error reading cache, using default preferences.",
+                    exception)
+                emit(emptyPreferences())
+              }
+              else -> throw exception
+            }
+          }
+          .map { preferences ->
+            val json = preferences[preferencesKey]
+            if (json != null) {
+              try {
+                val result =
+                    gson.fromJson(
+                        json, com.android.sample.model.preferences.Preferences::class.java)
```

Comment:
is it necessary to have this long name ? Would it compile by just puttin Preferences::class.java

Reply:
It's necessary because otherwise it would confuse it with the "Preferences" of Android's datastore which is not the same thing

Reply:
Ah yes indeed, again another proof that we really should have named it userPreferences